### ISSUE TITLE
SAML / Inbound / Trust Relationship: Hide Download Shibboleth Configuration if not "Generate" metadata

### ISSUE THREAD

**Issue by nynymike** [GluuFederation/oxTrust#331]
If the metadata is not provided via the generate method, then the link to "Download shibboleth metadata" should be hidden.

**Comment by aliaksander-samuseu** [comment]
Hi, @shekhar16 

Let me summarize it a bit for you.

We need to figure out 2 things:

1. Whether or not the link "Download Shibboleth Configuration" should be displayed for any TR other than created with "Generate" method. AFAIK this link is supposed to work with this method only.
2. What actually is the meaning of the fields which are displayed when "Generate" method is selected? What does "Url" field mean? What should be put in there? From the context, I can assume it should be entityID of the SP (for some other methods it will be url pointing to metadata, but "Generate" method is used when SP doesn't exist yet), in that case we must rename it accordingly, it will confuse a user. Then you should check why it rejects some valid urls (valid both as entityIDs and as urls itself), I tried `https://sphost-shib2.site:8443/shibboleth` and `https://sphost-shib2.site/shibboleth`, and in the end only after I left just these symbols - `https://sphostshib2.site`- it was accepted. And it also rejects valid certificate provided to it, in log there is a java exception like this (see full text in attachment):
```
2016-08-19 08:21:38,695 WARN [org.jboss.seam.ui.renderkit.DefaultActionRendererBase] Must set an id for the default action source 2016-08-19 08:22:03,612 ERROR [org.jboss.seam.exception.Exceptions] handled and logged exception javax.servlet.ServletException: java.lang.ClassCastException: org.bouncycastle.cert.X509CertificateHolder cannot be cast to java.security.cert.X509Certificate at javax.faces.webapp.FacesServlet.service(FacesServlet.java:606) at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303) at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208) at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) at 
```

I think you could ask @dmogn on your (dev) team for details, he is supposed to be our Shibboleth guru.

**Comment by aliaksander-samuseu** [comment]
Here is excerpt from our SAML TR docs:

_Generate: Using Gluu Server to generate configuration files for SP is another big option when the SP is inhouse application or “Shibboleth SP” is installed or going to be installed in target application site (SP). How to install Shibboleth SP will help user to configure and install Shibboleth SP on their own area. Please note few things when you are going to use Generate method for your SP._

- _URL: This is the hostname of SP_
- _Public certificate : You must have to provide the certificate which is Base64 encoded ASCII file and contain "-----BEGIN CERTIFICATE-----" and "-----END CERTIFICATE-----". This certificate can not be password protected._

_After creating the Trust Relationship, download the generated configuration files from Download Shibboleth2 configuration files link and place these configuration files inside your SP configuration._

So it seems "Url" should be a hostname, actually, while this field displays a hint in case of invalid value was provided, telling you that it must contain the scheme part too ("http://" or "https://"). It's a mess, we need to define the format of value it needs exactly, and set its name and hint's text accordingly. Though it would be more useful to have ability to provide exact entityID of the SP, and a custom port it listens at etc.

**Comment by shekhar16** [comment]
Fixed.
The link to "Download shibboleth metadata"  will be visible only for generate method,

**Comment by yurem** [comment]
"Public certificate" should be mandatory for generate method. Without this application  throws exception.

Even after selecting certificate we get:
2016-10-04 22:31:49,136 ERROR [org.jboss.seam.exception.Exceptions] handled and logged exception
javax.servlet.ServletException: java.lang.ClassCastException: org.bouncycastle.cert.X509CertificateHolder cannot be cast to java.security.cert.X509Certificate

**Comment by aliaksander-samuseu** [comment]
I've checked current state of this feature in 6-beta2 deb package. It's mostly fixed now (and design of controls on the page was changed, too), in case when "Generate" mode is chosen during initial creation  of TR, it behaves as expected, and in the end a working download link appears at the lower left corner of the page (didn't try to actually feed contents to Shibboleth).

But still there is some inconsistency here. If "File" method chosen initially instead, then this link appears as well, but now it doesn't work (the same "System error" message as before). Do we even expect it to be there, for this method of metadata provision? It assumes that SP already exists, so no need to generate Shib SP's configuration, right?

It happens because even when created using "Generate" method, the final method of metadata provision for this TR will be "File" anyway (this is what's being written to LDAP). Not sure why it's like this. May be we don't have a separate provision method called "Generate" atm, so have to resort to using a closest existing one? Having a separate "Generate" method with its own logic could probably help to get rid of this ambiguity.
